storage/mysql/root_store_test.go

```diff
@@ -0,0 +1,324 @@
+
+	type rootData struct {
+		roots      []*x509.Certificate
+		receivedAt time.Time
+	}
+
+	tests := []struct {
+		name             string
+		log              *ctlog.Log
+		rootsData        []rootData
+		wantErr          bool
+		wantRoots        []rootEntry
+		wantSets         []setEntry
+		wantObservations []obEntry
+	}{
+		{
+			name: "one root",
+			log:  pilot,
+			rootsData: []rootData{
+				{
+					roots: []*x509.Certificate{
+						root1,
+					},
+					receivedAt: aprilTimestamp,
+				},
+			},
+			wantRoots:        []rootEntry{e},
+			wantSets:         []setEntry{s},
+			wantObservations: []obEntry{o},
+		},
+		{
+			name: "duplicated root",
+			log:  pilot,
+			rootsData: []rootData{
+				{
+					roots: []*x509.Certificate{
+						root1,
+					},
+					receivedAt: aprilTimestamp,
+				},
+				{
+					roots: []*x509.Certificate{
+						root1,
+					},
+					receivedAt: mayTimestamp,
+				},
+			},
+			wantRoots:        []rootEntry{e},
+			wantSets:         []setEntry{s},
+			wantObservations: []obEntry{o, oMay},
+		},
+		{
+			name: "root order",
+			log:  pilot,
+			rootsData: []rootData{
+				{
+					roots: []*x509.Certificate{
+						root1,
+					},
+					receivedAt: aprilTimestamp,
+				},
+				{
+					roots: []*x509.Certificate{
+						root1, root2,
+					},
+					receivedAt: mayTimestamp,
+				},
+				{
+					roots: []*x509.Certificate{
+						root2, root1,
+					},
+					receivedAt: juneTimestamp,
+				},
+			},
+			wantRoots:        []rootEntry{e2, e},
+			wantSets:         []setEntry{s, s2, s1},
+			wantObservations: []obEntry{o, o2May, o2June},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+			testdb.Clean(ctx, testDB, "Roots")
+			testdb.Clean(ctx, testDB, "RootSets")
+			testdb.Clean(ctx, testDB, "RootSetObservations")
+			checkContents(ctx, t, nil, nil, nil)
+			st := NewRootStore(ctx, testDB)
+
+			var gotErr bool
+			for _, rdt := range test.rootsData {
+				if err := st.WriteRoots(ctx, test.log, rdt.roots, rdt.receivedAt); err != nil {
+					if !test.wantErr {
+						t.Fatalf("Storage.WriteRoots(ctx, %v, %v, %v) = %s, want nil", test.log, rdt.roots, rdt.receivedAt, err)
+					}
+					gotErr = true
+				}
+			}
+			if !gotErr && test.wantErr {
+				t.Fatal("Storage.WriteRoots() for all Root-sets produced no errors, wanted error")
+			}
+			checkContents(ctx, t, test.wantRoots, test.wantSets, test.wantObservations)
+		})
+	}
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if err := testdb.MySQLAvailable(); err != nil {
+		glog.Errorf("MySQL not available, skipping all MySQL storage tests: %v", err)
+		return
+	}
+	ctx := context.Background()
+	var err error
+	testDB, err = testdb.New(ctx, rootStoreSQL)
+	if err != nil {
+		glog.Exitf("failed to create test database: %v", err)
+	}
+	defer testDB.Close()
+	testdb.Clean(ctx, testDB, "Roots")
+	ec := m.Run()
+	os.Exit(ec)
+}
+
+var (
+	testDB       *sql.DB
+	rootStoreSQL = "rootStore.sql"
```

Comment:
This isn't the right filename anymore.
```suggestion
	rootStoreSQL = "root_store.sql"
```